### ISSUE TITLE
Support SSSD

### ISSUE THREAD

**Issue by fottsia** [ltsp/ltsp#338]
It would be great if LTSP could be deployed at sites that use Active Directory or LDAP for the user accounts. In that case, the LTSP image (or chrootless server) can just have SSSD configured:

https://wiki.ubuntu.com/Enterprise/Authentication/sssd

**Comment by alkisg** [comment]
LTSP5 supported LDAP by configuring the LTSP **server** as an LDAP **client**. This means that user listing (`getent passwd`) worked in the LTSP server, but not in the LTSP clients.
The LTSP5 display manager, LDM, didn't care if the user didn't exist. If the user could SSH to the server, LDM then used that connection to determine the user geometry (groups etc), and created that user on-the-fly in the client /etc/shadow while the user was logging in.

This is a big security hole, we don't want it in the new LTSP. So yes, the correct way to implement this is to make the LTSP **image** an LDAP **client**; to have user listing (`getent passwd`) working in LTSP clients before users log in.

With that approach, all that's left is for pamltsp to:
- Autodetect if the user comes from LDAP or passwd
- For LDAP users, it should completely skip SSH authentication (this is used when /home is NFS or locally mounted)
- But if SSHFS is needed, it should get the password from the PAM stack and pass it to SSHFS as usual

**Comment by jeremy-walker** [comment]
We currently have LTSP chroot clients that authenticate against an OpenLDAP server using SSSD on the clients. Our server is  running Ubuntu 20.04 headless and the clients are running Ubuntu Mate 20.04 with rootfs mounted over NFS. LightDM on the LTSP clients plays nicely with password policies implemented on OpenLDAP as well. Users can be forced to change passwords on login, and password expiry warnings work correctly and so do the many other constraints enforceable through password policies. 

In our setup, OpenLDAP runs on the same host as the LTSP server, but I'm don't think it would behave any differently with the LDAP server running on a different host. There is really nothing LTSP specific in the OpenLDAP server and SSSD client configurations. We mount /home via NFS4. The beauty of LTSP is that the SSSD configuration need only be done once on the chroot image, and all clients will automatically pick up the configuration/changes. 

One last thing, quoting from the link @fottsia posted:

> The option ldap_schema = rfc2307bis is related to Active Directory's LDAP schema, and is compatible with Windows Server 2003 R2 and newer. If you are using OpenLDAP or similar, ldap_schema = rfc2307 should be used instead. 

Not sure how accurate the second sentence is, since it says "rfc2307 should be used", not "can be used". We are using rfc2307bis (to enable support for memberOf overlay) with OpenLDAP and it works perfectly fine.

**Comment by alkisg** [comment]
I pushed an initial implementation and I uploaded a build in the LTSP proposed PPA.
Pamltsp now should support:

- The [aforementioned SSSD setup](https://wiki.ubuntu.com/Enterprise/Authentication/sssd)  (or similar LDAP setups). The administrator is supposed to configure the LTSP image or chroot to authenticate against LDAP, and pamltsp automatically handles the SSHFS mounting whenever that is needed (i.e. /home isn't already mounted via NFS).
- The usual SSH authentication for installations not using LDAP.
- Local accounts if /etc/shadow contains their password hashes.

There's no need to set any ltsp.conf variables, everything should be autodetected. Nevertheless, the following variables are respected:

- PAM_AUTH_TYPE="0|Primary|Additional": 0 means don't use pamltsp at all, e.g. for LDAP+NFS. "Primary" means do authentication with it, e.g. no LDAP. "Additional" means only do SSHFS with it, e.g. LDAP+SSHFS.
- SSH_SERVER="192.168.67.1": set a different SSH/SSHFS server than the boot server.
- SSH_OPTIONS="-p 8022": set a different SSH port or other SSH options.

I'll close this issue, but please test and report if I overlooked something.

**Comment by fottsia** [comment]
I tested LTSP with an LDAP server and an SSSD client and it worked great, except for one small detail.
When expiring the passwords on the LDAP server, the users were prompted to change them on the next logon. This worked fine with NFS /home.
But, when testing password renewals with SSHFS home, the user was able to change their password and login via SSSD, but a message was shown, "Pamltsp failed to mount home via SSHFS", and /home/username wasn't mounted.
After logoff/logon, /home/username was successfully mounted with SSHFS. It only fails the time when the password change is requested.

**Comment by alkisg** [comment]
Thanks for testing. While the correct thing to do here would be for pamltsp to fail the logon procedure, it can't, as when LDAP is detected, pamltsp gets an auxiliary role and it gets installed in the "Additional" section of /etc/pam.d/common-auth. Maybe there's some PAM setup that would allow pamltsp to be in the Primary section even in this case, but I haven't found it.
Another issue in this particular case is that /home/username needs to be created before the sshfs call.

So, the best thing that we can do, is to remove the empty /home/username directory, and then GUI logins will fail because they require $HOME to exist and be writable.
The users will get prompted for the password renewal, they'll change their passwords, pamltsp will notify them that home wasn't mounted, and then they'll have to login once more. Good enough for this edge case, I guess!

**Comment by carles-alemany** [comment]
Hi everyone, and thank you for this great project!

We've been using LTSP5 over the last 8 years, and now it's time to upgrade. On our old setup, which is still running, we have a Samba4 AD VM + LTSP5 VM serving different chroots to fat clients. LTSP server is joined to AD via pbis-open, fat-clients are able to authenticate with domain credentials and, every time a new user logs in, a new home directory is created into the LTSP server. 

We tried to replicate this setup with the new LTSP version, working with chroots as we did before. At this point, everything works, users are able to login with AD credentials but home folders are being created on the clients, not on the server, so everytime a client is disconnected, user's home directory is lost. 

**What we've done:** 

- Create a new chroot and assign a password to root.
- Export the new created chroot via NFS in order to boot from a client (https://github.com/ltsp/ltsp/wiki/chroots “Booting the chroot with a client”).
- Boot a client using the exported chroot.
- Join the chroot to AD with realmd-join, following these steps: https://computingforgeeks.com/join-ubuntu-debian-to-active-directory-ad-domain/
- After joining the chroot to AD, its name (focal) is being shown as a new computer inside the AD. 
- Shutdown the client
- Update chroot on the server (sudo update ltsp focal)

After doing these steps, a client can boot from the newly created image and we're able to login with a domain user and access SMB shares. Everything works except home directory, which is being created at the client instead of the server. 

Also, during the login, the client shows this message but then logs in without issues:
![image](https://user-images.githubusercontent.com/79309152/108489844-56b67a80-72a2-11eb-8ca3-ba48c8de0850.png)

Any ideas on how to automatically create user's home directory on the server at login? I'll be more than happy to make a wiki page if I can make it work. 

Thank you!
